torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+            orig_grad_data = param.grad.data
+
+            if self.gradient_predivide_factor > 1:
+                # Average grad by world_size for consistency with PyTorch DDP.
+                param.grad.div_(self.gradient_predivide_factor)
+
+            def _post_reduction_hook(obj) -> None:
+                """Hook to call on each param after the reduce-scatter."""
+                # If obj is a tensor, current stream is post_backward stream;
+                # If obj is a future, current stream is a new temporary team,
+                # use post_backward stream explicitly here, so that later on
+                # computation stream can wait for operations here
+                self._streams["post_backward"].wait_stream(torch.cuda.current_stream())
+                with torch.cuda.stream(self._streams["post_backward"]):
+                    assert param.grad is not None
+                    self._assert_state(TrainingState.BACKWARD_POST)
+                    if type(obj) == torch.Tensor:
+                        reduced_grad = obj
+                    else:
+                        reduced_grad = obj.value()[0]
+                    param.grad.data = reduced_grad
+                    if self.gradient_postdivide_factor > 1:
+                        # Average grad by world_size for consistency with PyTorch DDP.
+                        param.grad.div_(self.gradient_postdivide_factor)
+
+            if param._is_sharded:  # type: ignore[attr-defined]
+                assert param._is_sharded  # type: ignore[attr-defined]
+                grad_flatten = torch.flatten(param.grad)
+                chunks = list(grad_flatten.chunk(self.world_size))
+                num_pad = self.world_size * chunks[0].numel() - param.grad.numel()
+                input_flattened = F.pad(grad_flatten, [0, num_pad])
+                output = torch.zeros_like(chunks[0])
+                dist._reduce_scatter_base(
+                    output,
+                    input_flattened,
+                    group=self.process_group,
+                    async_op=True).get_future().then(_post_reduction_hook)
+            else:
+                # Currently the only way for _is_sharded to be False is if
+                # world_size == 1. This could be relaxed in the future, e.g,
+                # no sharding like PyTorch DDP, in which case grads should be
+                # all-reduced here.
+                assert self.world_size == 1
+                _post_reduction_hook(param.grad.data)
+
+            # After _post_backward_hook returns, orig_grad_data will eventually
+            # go out of scope, at which point it could otherwise be freed for
+            # further reuse by the main stream while the div/reduce_scatter/copy
+            # are underway in the post_backward stream. See:
+            # github.com/NVIDIA/apex/blob/master/apex/parallel/distributed.py
+            orig_grad_data.record_stream(self._streams["post_backward"])
```

Comment:
@albanD would I be correct if I assume this won't guard the comp op in the future callback, as those are not inserted in the streams yet?

Reply:
I'm not sure what the idea here is tbh.
This function is making sure that all ops on this stream are done before this Tensor can be deleted.
I'm not sure what is the problem you're talking about? If the Tensor goes out of scope, nothing can use it anymore right?

Reply:
The concern is that when `_post_reduction_hook` is used as a future callback, more CUDA ops on `orig_grad_data` can be inserted into the `post_backward` stream after this point. Basically, when using a future, the comment above this line might no longer be true, as `_post_backward_hook` might not return yet. 


> After _post_backward_hook returns, orig_grad_data will eventually
> go out of scope, at which point it could otherwise be freed for
> further reuse by the main stream while the div/reduce_scatter/copy
> are underway in the post_backward stream.

Reply:
Had second thought, actually we can change this reduce_scatter back to be blocking. It is OK as it is scheduled under "post_backward" stream and will be blocking in this streams as well. That being said, it will not block computation stream and can have overlap with computation if we change it to be blocking. I will do a benchmark to make sure the performance is the same after changing it to be blocking op. 

I'am actually concerned about using future here: 
1. as you mentioned, the orig_grad_data can not be guaranteed to be kept in memory
2. I noticed that future did not throw exception when there is error in the post call back, the silent error caused incorrect results and hide the root cause when I ran tests